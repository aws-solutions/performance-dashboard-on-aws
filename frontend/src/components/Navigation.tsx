/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect } from "react";
import { NavHashLink } from "react-router-hash-link";

interface WidgetNameId {
    name: string;
    id: string;
    isInsideSection: boolean;
    sectionWithTabs: string;
}

interface Props {
    stickyPosition: number;
    offset: number;
    widgetNameIds: Array<WidgetNameId>;
    activeWidgetId: string;
    onBottomOfThePage: Function;
    isTop: boolean;
    area: number;
    marginRight: number;
    displayTableOfContents: boolean;
    onClick?: Function;
}

function Navigation({
    stickyPosition,
    offset,
    widgetNameIds,
    activeWidgetId,
    onBottomOfThePage,
    isTop,
    area,
    displayTableOfContents,
    onClick,
    marginRight,
}: Props) {
    // forcefully highlight the last tab in table of contents when user reaches
    // the bottom of the page
    const handleScroll = () => {
        const isBottom =
            Math.ceil(window.innerHeight + window.scrollY) >= document.documentElement.scrollHeight;
        if (isBottom && widgetNameIds.length) {
            onBottomOfThePage(widgetNameIds[widgetNameIds.length - 1].id);
        }
    };

    useEffect(() => {
        window.addEventListener("scroll", handleScroll, {
            passive: true,
        });

        return () => {
            window.removeEventListener("scroll", handleScroll);
        };
    }, []);

    const scrollWithOffset = (el: HTMLElement) => {
        const yCoordinate = el.getBoundingClientRect().top + window.scrollY;
        const yOffset = -offset;
        window.scrollTo({ top: yCoordinate + yOffset, behavior: "smooth" });
    };

    const onClickHandler = (active: string) => {
        if (onClick) {
            onClick(active);
        }
    };

    if (!isTop) {
        return (
            <div
                className={`tablet:grid-col-${area}`}
                style={{
                    position: "sticky",
                    top: stickyPosition,
                    float: "right",
                    marginRight: `-${marginRight}%`,
                    borderLeft: "1px solid rgba(200, 200, 200, 1)",
                }}
                hidden={!displayTableOfContents}
            >
                <nav aria-label="Secondary navigation">
                    <ul
                        className="usa-sidenav"
                        style={{
                            fontSize: "0.9rem",
                            borderStyle: "none",
                            margin: "3.5em 0 3.5em 0",
                        }}
                    >
                        {widgetNameIds.map((widget) => {
                            return (
                                <li
                                    className={`usa-sidenav__item ${
                                        widget.id === activeWidgetId ? "usa-current" : ""
                                    }`}
                                    key={widget.id + "link"}
                                    style={{
                                        borderStyle: "none",
                                    }}
                                    onClick={() => onClickHandler(widget.id)}
                                >
                                    <NavHashLink
                                        to={
                                            "#" +
                                            (widget.isInsideSection && widget.sectionWithTabs
                                                ? widget.sectionWithTabs
                                                : widget.id)
                                        }
                                        scroll={(el) => scrollWithOffset(el)}
                                        style={{
                                            paddingLeft: widget.isInsideSection ? "32px" : "10px",
                                        }}
                                    >
                                        {widget.name}
                                    </NavHashLink>
                                </li>
                            );
                        })}
                    </ul>
                </nav>
            </div>
        );
    } else {
        return (
            <nav aria-label="Secondary navigation" hidden={!displayTableOfContents}>
                <ul
                    className="usa-sidenav"
                    style={{
                        fontSize: "0.9rem",
                        borderStyle: "none",
                        marginLeft: 0,
                    }}
                >
                    {widgetNameIds.map((widget) => {
                        return (
                            <li
                                className={`usa-sidenav__item ${
                                    widget.id === activeWidgetId ? "usa-current" : ""
                                }`}
                                key={widget.id + "link"}
                                style={{
                                    borderStyle: "none",
                                }}
                                onClick={() => onClickHandler(widget.id)}
                            >
                                <NavHashLink
                                    to={
                                        "#" +
                                        (widget.isInsideSection && widget.sectionWithTabs
                                            ? widget.sectionWithTabs
                                            : widget.id)
                                    }
                                    scroll={(el) => scrollWithOffset(el)}
                                    style={{
                                        paddingLeft: widget.isInsideSection ? "36px" : "16px",
                                    }}
                                >
                                    {widget.name}
                                </NavHashLink>
                            </li>
                        );
                    })}
                </ul>
            </nav>
        );
    }
}

export default Navigation;
