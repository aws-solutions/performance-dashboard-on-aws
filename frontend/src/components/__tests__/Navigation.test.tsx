/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { render, screen, fireEvent } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import Navigation from "../Navigation";

describe("Navigation", () => {
    const props = {
        stickyPosition: 0,
        offset: 0,
        widgetNameIds: [
            {
                name: "first widget",
                id: "uuid",
                isInsideSection: false,
                sectionWithTabs: "",
            },
            {
                name: "second widget",
                id: "uuid-2",
                isInsideSection: false,
                sectionWithTabs: "",
            },
        ],
        activeWidgetId: "uuid",
        onBottomOfThePage: jest.fn(),
        isTop: false,
        area: 2,
        marginRight: 0,
        displayTableOfContents: true,
        onClick: jest.fn(),
    };
    test("renders the component", async () => {
        const wrapper = render(
            <Navigation
                stickyPosition={props.stickyPosition}
                offset={props.offset}
                area={props.area}
                marginRight={props.marginRight}
                widgetNameIds={props.widgetNameIds}
                activeWidgetId={props.activeWidgetId}
                isTop={props.isTop}
                displayTableOfContents={props.displayTableOfContents}
                onBottomOfThePage={props.onBottomOfThePage}
                onClick={props.onClick}
            />,
            { wrapper: MemoryRouter },
        );
        expect(wrapper.container).toMatchSnapshot();
    });

    test("renders the component at the top", async () => {
        const wrapper = render(
            <Navigation
                stickyPosition={props.stickyPosition}
                offset={props.offset}
                area={props.area}
                marginRight={props.marginRight}
                widgetNameIds={props.widgetNameIds}
                activeWidgetId={props.activeWidgetId}
                isTop={true}
                displayTableOfContents={props.displayTableOfContents}
                onBottomOfThePage={props.onBottomOfThePage}
                onClick={props.onClick}
            />,
            { wrapper: MemoryRouter },
        );
        expect(wrapper.container).toMatchSnapshot();
    });

    test("click on a navigation link should trigger onClick callback", async () => {
        render(
            <>
                <Navigation
                    stickyPosition={props.stickyPosition}
                    offset={props.offset}
                    area={props.area}
                    marginRight={props.marginRight}
                    widgetNameIds={props.widgetNameIds}
                    activeWidgetId={props.activeWidgetId}
                    isTop={props.isTop}
                    displayTableOfContents={props.displayTableOfContents}
                    onBottomOfThePage={props.onBottomOfThePage}
                    onClick={props.onClick}
                />
                {props.widgetNameIds.map((widget) => (
                    <div id={widget.id} key={widget.id}>
                        {widget.name}
                    </div>
                ))}
            </>,
            { wrapper: MemoryRouter },
        );
        const navigationLink = screen.getByRole("link", {
            name: "second widget",
        });
        fireEvent.click(navigationLink);

        expect(props.onClick).toBeCalled();
    });

    test("click on the navigation link should trigger onClick callback when component at the top", async () => {
        render(
            <>
                <Navigation
                    stickyPosition={props.stickyPosition}
                    offset={props.offset}
                    area={props.area}
                    marginRight={props.marginRight}
                    widgetNameIds={props.widgetNameIds}
                    activeWidgetId={props.activeWidgetId}
                    isTop={true}
                    displayTableOfContents={props.displayTableOfContents}
                    onBottomOfThePage={props.onBottomOfThePage}
                    onClick={props.onClick}
                />
                {props.widgetNameIds.map((widget) => (
                    <div id={widget.id} key={widget.id}>
                        {widget.name}
                    </div>
                ))}
            </>,
            { wrapper: MemoryRouter },
        );
        const navigationLink = screen.getByRole("link", {
            name: "second widget",
        });
        fireEvent.click(navigationLink);

        expect(props.onClick).toBeCalled();
    });

    test("scroll to bottom should trigger onBottomOfThePage callback", async () => {
        render(
            <Navigation
                stickyPosition={props.stickyPosition}
                offset={props.offset}
                area={props.area}
                marginRight={props.marginRight}
                widgetNameIds={props.widgetNameIds}
                activeWidgetId={props.activeWidgetId}
                isTop={props.isTop}
                displayTableOfContents={props.displayTableOfContents}
                onBottomOfThePage={props.onBottomOfThePage}
                onClick={props.onClick}
            />,

            { wrapper: MemoryRouter },
        );
        fireEvent.scroll(window, {
            target: { scrollY: 100 },
        });
        expect(props.onBottomOfThePage).toBeCalled();
    });
});
