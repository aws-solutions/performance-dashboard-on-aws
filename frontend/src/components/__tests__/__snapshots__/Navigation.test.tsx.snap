// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders the component 1`] = `
<div>
  <div
    class="tablet:grid-col-2"
    style="position: sticky; top: 0px; float: right; margin-right: -0%; border-left: 1px solid rgba(200, 200, 200, 1);"
  >
    <nav
      aria-label="Secondary navigation"
    >
      <ul
        class="usa-sidenav"
        style="font-size: 0.9rem; border-style: none; margin: 3.5em 0px 3.5em 0px;"
      >
        <li
          class="usa-sidenav__item usa-current"
          style="border-style: none;"
        >
          <a
            href="/#uuid"
            style="padding-left: 10px;"
          >
            first widget
          </a>
        </li>
        <li
          class="usa-sidenav__item "
          style="border-style: none;"
        >
          <a
            href="/#uuid-2"
            style="padding-left: 10px;"
          >
            second widget
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`Navigation renders the component at the top 1`] = `
<div>
  <nav
    aria-label="Secondary navigation"
  >
    <ul
      class="usa-sidenav"
      style="font-size: 0.9rem; border-style: none; margin-left: 0px;"
    >
      <li
        class="usa-sidenav__item usa-current"
        style="border-style: none;"
      >
        <a
          href="/#uuid"
          style="padding-left: 16px;"
        >
          first widget
        </a>
      </li>
      <li
        class="usa-sidenav__item "
        style="border-style: none;"
      >
        <a
          href="/#uuid-2"
          style="padding-left: 16px;"
        >
          second widget
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
